### ISSUE TITLE
[BUG] Retrieved chunks from RAG pipeline (source_documents) not being displayed correctly

### ISSUE THREAD

**Issue by ai-learner-00** [mlflow/mlflow#11171]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

- Client: 2.10.2
- Tracking server: 2.10.2

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Python version**: 3.10.13
- **yarn version, if running the dev UI**:

### Describe the problem

I was wondering if I could see the retrieved chunks from a RAG pipeline.

```python
qa = RetrievalQA.from_chain_type(
    llm=OpenAI(temperature=0.7),
    chain_type="stuff",
    retriever=docsearch.as_retriever(),
    return_source_documents=True,
)
eval_df = pd.DataFrame(
    {
        "questions": [
            "What is MLflow?",
            "How to run mlflow.evaluate()?",
            "How to log_table()?",
            "How to load_table()?",
        ],
    }
)
```

However, I see "object [Object]". Is this intended/expected?

![image](https://github.com/mlflow/mlflow/assets/158079614/292639bb-1c31-4738-92e2-22eab525bdff)

If I look at the results from `results = mlflow.evaluate()`, I do see the retrieved chunk.

```python
results.tables["eval_results_table"].iloc[0]["source_documents"] 
```
```
[{'lc_attributes': {},
  'lc_secrets': {},
  'metadata': {'language': 'en',
   'source': 'https://mlflow.org/docs/latest/index.html',
   'title': 'MLflow: A Tool for Managing the Machine Learning Lifecycle — MLflow 2.10.2 documentation'},
  'page_content': 'MLflow: A Tool for Managing the Machine Learning Lifecycle \nMLflow is an open-source platform, purpose-built to assist machine learning practitioners and teams in\nhandling the complexities of the machine learning process. MLflow focuses on the full lifecycle for\nmachine learning projects, ensuring that each phase is manageable, traceable, and reproducible.\nIn each of the sections below, you will find overviews, guides, and step-by-step tutorials to walk you through\nthe features of MLflow and how they can be leveraged to solve real-world MLOps problems.\n\nGetting Started with MLflow \nIf this is your first time exploring MLflow, the tutorials and guides here are a great place to start. The emphasis in each of these is\ngetting you up to speed as quickly as possible with the basic functionality, terms, APIs, and general best practices of using MLflow in order to\nenhance your learning in area-specific guides and tutorials.\n\nGetting Started Guides and Quickstarts',
  'type': 'Document'},
 {'lc_attributes': {},
  'lc_secrets': {},
  'metadata': {'language': 'en',
   'source': 'https://mlflow.org/docs/latest/index.html',
   'title': 'MLflow: A Tool for Managing the Machine Learning Lifecycle — MLflow 2.10.2 documentation'},
  'page_content': 'MLflow: A Tool for Managing the Machine Learning Lifecycle — MLflow 2.10.2 documentation\n\n2.10.2\n\n\n MLflow\n\nWhat is MLflow?\nGetting Started with MLflow\nNew Features\nLLMs\nModel Evaluation\nDeep Learning\nTraditional ML\nDeployment\nMLflow Tracking\nSystem Metrics\nMLflow Projects\nMLflow Models\nMLflow Model Registry\nMLflow Recipes\nMLflow Plugins\nMLflow Authentication\nCommand-Line Interface\nSearch Runs\nSearch Experiments\nPython API\nR API\nJava API\nREST API\nOfficial MLflow Docker Image\nCommunity Model Flavors\nTutorials and Examples\n\n\nContribute\n\n\nDocumentation \nMLflow: A Tool for Managing the Machine Learning Lifecycle',
  'type': 'Document'},
 {'lc_attributes': {},
  'lc_secrets': {},
  'metadata': {'language': 'en',
   'source': 'https://mlflow.org/docs/latest/index.html',
   'title': 'MLflow: A Tool for Managing the Machine Learning Lifecycle — MLflow 2.10.2 documentation'},
  'page_content': 'Deep Learning \nSee how MLflow can help manage the full lifecycle of your Deep Learning projects. Whether you’re using frameworks like\nTensorFlow (tensorflow), Keras (keras), PyTorch (torch), Fastai (fastai), or spaCy (spacy), MLflow offers first-class support,\nensuring seamless integration and deployment. Additionally, generic packaging frameworks that have native MLflow integration such as ONNX (onnx)\ngrealy help to simplify the deployment of deep learning models to a wide variety of deployment providers and environments.\nPaired with MLflow’s streamlined APIs and comparative UI, you are equipped with everything needed to manage, track, and optimize your deep learning workflows.',
  'type': 'Document'},
 {'lc_attributes': {},
  'lc_secrets': {},
  'metadata': {'language': 'en',
   'source': 'https://mlflow.org/docs/latest/index.html',
   'title': 'MLflow: A Tool for Managing the Machine Learning Lifecycle — MLflow 2.10.2 documentation'},
  'page_content': 'Traditional ML \nLeverage the power of MLflow for all your Traditional Machine Learning needs. Whether you’re working with supervised, unsupervised, statistical, or time series data,\nMLflow streamlines the process by providing an integrated environment that supports a large array of widely-used libraries like Scikit-learn (sklearn),\nSparkML (spark), XGBoost (xgboost), LightGBM (lightgbm), CatBoost (catboost), Statsmodels, Prophet, and Pmdarima.\nWith MLflow, you not only get APIs tailored for these libraries but also a user-friendly UI to compare various runs, ensuring that your model tuning and\nevaluation phases are both efficient and insightful.\n\nTraditional ML Guides and Tutorials \n\n                    Hyperparameter Tuning with Optuna and MLflow',
  'type': 'Document'}]
```

---

Also, I am confused by how the faithfulness metric uses the retrieved chunks.

```python
def model(input_df):
    answer = []
    for index, row in input_df.iterrows():
        answer.append(qa(row["questions"])) # passing the question only, not the input_df

    return answer

results = mlflow.evaluate(
    model,
    eval_df,
    model_type="question-answering",
    evaluators="default",
    predictions="result",
    extra_metrics=[faithfulness_metric,
                   relevance_metric, mlflow.metrics.latency()],
    evaluator_config={
        "col_mapping": {
            "inputs": "questions",
            "context": "source_documents", # context
        }
    },
```

The input dataframe doesn't have a `source_documents` column. Is it looking for it in the answer instead of the input dataframe?
```python
def model(input_df):
    answers = []
    for index, row in input_df.iterrows():
        answer = qa(row["questions"])
        answers.append(answer)
        print(f"\nANSWER:\n{answer}\n")

    return answers
```
```
ANSWER:
{'query': 'What is MLflow?', 'result': ' MLflow is an open-source platform designed to assist machine learning practitioners and teams in managing the complexities of the machine learning process. It focuses on the full lifecycle of machine learning projects, ensuring that each phase is manageable, traceable, and reproducible. ', 'source_documents': [Document(page_content='MLflow: A Tool for Managing the Machine Learning Lifecycle \nMLflow is 
...
```

 Is the fact the `context` isn't pure text detected (and `page_content` extracted)?

### Tracking information

```shell
System information: Windows 10.0.19045
Python version: 3.10.13
MLflow version: 2.10.2
MLflow module location: c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\__init__.py
Tracking URI: http://127.0.0.1:5000
Registry URI: http://127.0.0.1:5000
MLflow dependencies: 
  Flask: 3.0.1
  Jinja2: 3.1.3
  aiohttp: 3.9.2
  alembic: 1.13.1
  boto3: 1.34.37
  botocore: 1.34.37
  click: 8.1.7
  cloudpickle: 3.0.0
  docker: 7.0.0
  entrypoints: 0.4
  fastapi: 0.109.0
  gitpython: 3.1.41
  importlib-metadata: 6.11.0
  kubernetes: 29.0.0
  markdown: 3.5.2
  matplotlib: 3.8.2
  numpy: 1.26.3
  packaging: 23.2
  pandas: 2.2.0
  protobuf: 4.23.4
  pyarrow: 15.0.0
  pydantic: 2.6.0
  pytz: 2023.4
  pyyaml: 6.0.1
  querystring-parser: 1.2.4
  requests: 2.31.0
  scikit-learn: 1.4.0
  scipy: 1.12.0
  slowapi: 0.1.8
  sqlalchemy: 2.0.25
  sqlparse: 0.4.4
  tiktoken: 0.5.2
  uvicorn: 0.27.0.post1
  waitress: 2.1.2
  watchfiles: 0.21.0
```

### Code to reproduce issue

```python
import pandas as pd
import mlflow
from langchain.chains import RetrievalQA
from langchain.document_loaders import WebBaseLoader
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.llms import OpenAI
from langchain.text_splitter import CharacterTextSplitter
from langchain.vectorstores import Chroma

loader = WebBaseLoader("https://mlflow.org/docs/latest/index.html")

documents = loader.load()
text_splitter = CharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
texts = text_splitter.split_documents(documents)

embeddings = OpenAIEmbeddings()
docsearch = Chroma.from_documents(texts, embeddings)

qa = RetrievalQA.from_chain_type(
    llm=OpenAI(temperature=0.7),
    chain_type="stuff",
    retriever=docsearch.as_retriever(),
    return_source_documents=True,
)

def model(input_df):
    answer = []
    for index, row in input_df.iterrows():
        answer.append(qa(row["questions"]))

    return answer


eval_df = pd.DataFrame(
    {
        "questions": [
            "What is MLflow?",
            "How to run mlflow.evaluate()?",
            "How to log_table()?",
            "How to load_table()?",
        ],
    }
)

from mlflow.metrics.genai import faithfulness, relevance

faithfulness_metric = faithfulness(model="openai:/gpt-4")
relevance_metric = relevance(model="openai:/gpt-4")

mlflow.set_tracking_uri("http://127.0.0.1:5000")
mlflow.set_experiment("Evaluate RAG")

results = mlflow.evaluate(
    model,
    eval_df,
    model_type="question-answering",
    evaluators="default",
    predictions="result",
    extra_metrics=[faithfulness_metric,
                   relevance_metric, mlflow.metrics.latency()],
    evaluator_config={
        "col_mapping": {
            "inputs": "questions",
            "context": "source_documents",
        }
    },
)
results.tables["eval_results_table"].iloc[0]["source_documents"]
```

### Stack trace

```
2024/02/16 13:16:56 INFO mlflow.tracking.fluent: Experiment with name 'Evaluate RAG' does not exist. Creating a new experiment.
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\data\digest_utils.py:29: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_columns = trimmed_df.columns[(df.applymap(type) == str).all(0)]
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\models\evaluation\base.py:414: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
2024/02/16 13:16:57 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/02/16 13:16:57 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\langchain_core\_api\deprecation.py:117: LangChainDeprecationWarning: The function `__call__` was deprecated in LangChain 0.1.0 and will be removed in 0.2.0. Use invoke instead.
  warn_deprecated(
2024/02/16 13:17:01 INFO mlflow.models.evaluation.default_evaluator: Testing metrics on first row...
WARNING:tensorflow:From c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\keras\src\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.

Using default facebook/roberta-hate-speech-dynabench-r4-target checkpoint
100%
1/1 [00:03<00:00, 3.82s/it]
100%
1/1 [00:05<00:00, 5.02s/it]
2024/02/16 13:17:23 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: token_count
2024/02/16 13:17:23 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: toxicity
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: flesch_kincaid_grade_level
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: ari_grade_level
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: exact_match
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating metrics: faithfulness
100%
4/4 [00:08<00:00, 1.62s/it]
2024/02/16 13:17:33 INFO mlflow.models.evaluation.default_evaluator: Evaluating metrics: relevance
100%
4/4 [00:06<00:00, 1.29s/it]
```

### Other info / logs

```
2024/02/16 13:16:56 INFO mlflow.tracking.fluent: Experiment with name 'Evaluate RAG' does not exist. Creating a new experiment.
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\data\digest_utils.py:29: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_columns = trimmed_df.columns[(df.applymap(type) == str).all(0)]
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\models\evaluation\base.py:414: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
2024/02/16 13:16:57 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/02/16 13:16:57 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\langchain_core\_api\deprecation.py:117: LangChainDeprecationWarning: The function `__call__` was deprecated in LangChain 0.1.0 and will be removed in 0.2.0. Use invoke instead.
  warn_deprecated(
2024/02/16 13:17:01 INFO mlflow.models.evaluation.default_evaluator: Testing metrics on first row...
WARNING:tensorflow:From c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\keras\src\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.

Using default facebook/roberta-hate-speech-dynabench-r4-target checkpoint
100%
1/1 [00:03<00:00, 3.82s/it]
100%
1/1 [00:05<00:00, 5.02s/it]
2024/02/16 13:17:23 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: token_count
2024/02/16 13:17:23 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: toxicity
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: flesch_kincaid_grade_level
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: ari_grade_level
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating builtin metrics: exact_match
2024/02/16 13:17:24 INFO mlflow.models.evaluation.default_evaluator: Evaluating metrics: faithfulness
100%
4/4 [00:08<00:00, 1.62s/it]
2024/02/16 13:17:33 INFO mlflow.models.evaluation.default_evaluator: Evaluating metrics: relevance
100%
4/4 [00:06<00:00, 1.29s/it]
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [X] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by ai-learner-00** [comment]
To reformulate, when the `context`/`retrieved_documents` is injected into the prompt, it needs to be a string, so I was expecting a full string on the UI as well.

**Comment by kalanyuz** [comment]
@ai-learner-00 
if you need to see something fast, I did the following in the model function:
```python
def model(input_df):
    answer = []
    for index, row in input_df.iterrows():
        # mine was ConversationalRetrievalChain, but you can keep yours as is
        answer.append(qa({"question":row["questions"], "chat_history": row["chat_history"]}))
        answer[-1]['source_documents'] = list(map(lambda x: x.json(), answer[-1]['source_documents']))```
    
    return answer
